### ISSUE TITLE
Installation fail on Ubuntu 20.04 LTS armhf if package whiptail is not installed

### ISSUE THREAD

**Issue by crisrc012** [pi-hole/pi-hole#3297]
**In raising this issue, I confirm the following:**

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*.
- [X] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

5

---
**Expected behaviour:**

Prerequisite validation should install the package `whiptail` if it is needed.

**Actual behaviour:**

Exit with the message: `Cancel was selected, exiting installer`
![imagen](https://user-images.githubusercontent.com/10794035/80175277-61d14900-85b2-11ea-9fd5-8998f397350b.png)

**Steps to reproduce:**

1. Uninstall the package `whiptail`: `sudo apt remove -y whiptail`.
2. Try to install pi-hole with the instalation guide.

**Workaround**

* Manually install `whiptail` package: `sudo apt install -y whiptail`
* Install pi-hole with the installation guide

---
`NOTE: English is not my native language.`

**Comment by dschaper** [comment]
It's already a required package. https://github.com/pi-hole/pi-hole/blob/master/automated%20install/basic-install.sh#L246

Your output is from a bash session with `set -x`?

**Comment by crisrc012** [comment]
I was trying to install using `sudo bash basic-install.sh` and the output was `Cancel was selected, exiting installer` then I executed `sudo bash -xv basic-install.sh`, and I notice that `whiptail` should be installed.
I manually installed `whiptail` to resolve the issue.

**Comment by MichaIng** [comment]
@dschaper @crisrc012 
Remember that this has been fixed with #2962. Please try with the current Pi-hole v5 beta.

**Comment by MichaIng** [comment]
@crisrc012 
Although the solution is quite easy, the issue that the installer calls `whiptail` (in certain circumstances) before it gets installed is still present. The case is obviously rare, but worth to keep the issue open so someone could think of a failsafe solution when finding time?

**Comment by juliangaal** [comment]
Can confirm that this issue happens on 20.04 x86 as well

**Comment by MichaIng** [comment]
@juliangaal 
How exactly did you face it, fresh install or update, and if so from which previous Pi-hole version?
- On a fresh Pi-hole install, `whiptail` is installed now before it is called the first time, so that should not be an issue anymore.
- On `pihole -up`, there are still `whiptail` calls done before installer dependencies are installed, but on initial install `whiptail` should have been installed, so when one does not manually remove it, it should not be possible to run into such a situation with Pi-hole updates.

**Comment by juliangaal** [comment]
I manually deleted all pihole dependencies for some tests (including whiptail). Then tried to reinstall. I realize this isn't the standard route, sry for giving you a tiny heart attack...

**Comment by MichaIng** [comment]
> sry for giving you a tiny heart attack...

😄 don't worry, all fine. I think it is still an issue that installer dependencies are installed after some possible whiptail prompts on updates, but it's a minor one as you stated. It's unlikely and can be even seen as own fault when packages, installed/verified as part of the Pi-hole installer, are manually removed while Pi-hole is still used 😛.

**Comment by juliangaal** [comment]
ok, glad you're fine :D have a good one

**Comment by yubiuser** [comment]
I was going to work on the issue, but discovered that @PromoFaux already solved it. I'm sure it is a 100% intended side-effect of https://github.com/pi-hole/pi-hole/pull/4260 ;-)

As you can see here
https://github.com/pi-hole/pi-hole/blob/release/v5.4/automated%20install/basic-install.sh#L2563

In the current beta branch, `istall_dependent_packages "${OS_CHECK_DEPS[@]}` and `install_dependent_packages "${INSTALLER_DEPS[@]}"` are called before `update_dialogs`.

Going to mark this as fixed in the next release

**Comment by MichaIng** [comment]
Mostly, indeed. There are still two `whiptail` calls in the RPM block of `package_manager_detect`. Also, I just recognised that the PR has not been merged into the development branch, but only the 5.4 release branch 🤔. To not un-resolve things with a next version, this should be done quickly to not run into merge conflicts 😉.

**Comment by yubiuser** [comment]
I think it got a bit out of order....

At a certain point in time it must be release/v5.4 -> development -> master
___

Thanks for the close look... I just noticed `whiptail` is missing in the RPM block also in the `INSTALLER_DEPS`....

**Comment by MichaIng** [comment]
Ah true, probably it is on RPM distros what on APT/DPkg distros an "essential" package is, so practically always installed, unless someone intentionally confirms to destroy its system 😄.

What I remember is that the package is not called "whiptail" there, but "newt", the underlying library.

**Comment by yubiuser** [comment]
Good you remembered, I didn't know at all.

So we have the two whiptails left, but also two instances of grep, which is also part of OS_CHECK_DEPS

**Comment by MichaIng** [comment]
At least on Debian, [grep](https://packages.debian.org/buster/grep) is an "essential" package:
- It cannot be removed unless you manually enter a sentence that you know what you are doing:
    ```
    The following packages will be REMOVED:
      grep*
    WARNING: The following essential packages will be removed.
    This should NOT be done unless you know exactly what you are doing!
      grep
    0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
    After this operation, 1117 kB disk space will be freed.
    You are about to do something potentially harmful.
    To continue type in the phrase 'Yes, do as I say!'
     ?]
    ```
- Furthermore, after doing so, every `apt-get install ...` and `apt-get upgrade` will automatically reinstall `grep` as well.
- To keep it uninstalled, it is required to add an APT pin `-1` for that package, to force APT not seeing it as valid candidate.
- Essential packages are never required to be APT dependencies in the repository, so many packages (or their postinst scripts) use essential tools like grep, sed, gzip, kill, ls, ... and do not need to have the providing package added as dependency. So many things will break when these commands are not present.
- So basically unless someone very intentionally breaks its system, there is no need to check for any of such essential packages being present or not, at least on on Debian and variants. Not sure how Fedora/RPM distros handle this.

**Comment by yubiuser** [comment]
Fixed in:

https://discourse.pi-hole.net/t/pi-hole-ftl-v5-9-web-v5-6-and-core-v5-4-released/49544/1
and
https://github.com/pi-hole/pi-hole/releases/tag/v5.5
